jxl/src/api/output.rs

```diff
@@ -5,95 +5,169 @@
 
 #![allow(unsafe_code)]
 
-use std::{mem::MaybeUninit, ops::DerefMut};
+use core::slice;
+use std::{marker::PhantomData, mem::MaybeUninit, ops::Range};
 
-/// Trait to represent image output buffers.
-/// This trait is implemented for
-/// - [u8], which will be interpreted as contiguous rows of data, starting from row 0 and
-///   followed by row 1, 2, ... without any gaps)
-/// - [&mut [u8]], which will be interpreted as buffers for distinct rows (starting from
-///   the buffer for row 0, then row 1, ...).
-pub trait JxlOutputBuffer<'a> {
-    /// Returns buffers for each row of the image. Buffers might be uninitialized, but the caller
-    /// guarantees that data that *is* initialized in the buffers is not overwritten with
-    /// uninitialized data.
-    ///
-    /// # Safety
-    /// The returned buffers must not be populated with uninit data.
-    unsafe fn get_row_buffers(
-        &'a mut self,
-        shape: (usize, usize),
-        bytes_per_pixel: usize,
-    ) -> impl DerefMut<Target = [&'a mut [MaybeUninit<u8>]]>;
+use num_traits::ToBytes;
+
+use crate::image::Image;
+
+pub struct JxlOutputBuffer<'a> {
+    // Safety invariants:
+    //  - uninit data is never written to `buf`.
+    //  - `buf` is valid for writes for all bytes in the range
```

Comment:
What is 'valid' in this context? Isn't all this just saying that 'buf' must have at least size E (as defined below)?

Reply:
It's more complicated than that, see https://doc.rust-lang.org/std/ptr/index.html#safety